### ISSUE TITLE
ES6 Generator support

### ISSUE THREAD

**Issue by wizzard0** [Microsoft/TypeScript#1564]
This can be broken down to 3 areas:
1. basic (parser) support, even if we lose typechecking locally
2. Typechecking support
3. Transpiling like facebook/regenerator or Traceur, for targeting ES3/ES5.

Any timelines on this?

**Comment by JsonFreeman** [comment]
Please take a look at #2873 and let me know if this addresses your needs.

**Comment by buzinas** [comment]
I've just seen that the TypeScript 2.0 milestone was added to support transpiling generators to ES5: do you have any estimate when it will be released?

**Comment by omidkrad** [comment]
@buzinas I don't know when to expect it but until then the workaround is to get TypeScript code with async/await transpiled to ES6 and then in a second build step use [Babel](https://babeljs.io) to transpile that to ES5.

**Comment by buzinas** [comment]
@omidkrad Yeah, I know that, did this for one project, but for the others, I'm trying to keep things dry, so, I'm using only TypeScript as transpiler.

For bigger projects, the build time of two transformations doesn't compensate for us. Also, the sourcemaps don't work well when transpiling two times, unfortunately (they work, but poorly sometimes).

At the moment, we're sticking with promises, and it's good enough for 90% of the cases.

**Comment by DanielRosenwasser** [comment]
I think we'll have a beta ready to use towards late May/early July, though no promises on dates.

**Comment by piotrwitek** [comment]
@omidkrad For me using Babel just for transpiling async/await to ES5 was a a little bit of overkill, so for that workaround in my projects I'm currently using https://github.com/facebook/regenerator, works fine, but I cannot wait to finally have it possible by only using TS

**Comment by omidkrad** [comment]
@piotrwitek Nice tip! Do you run generator on tsc's ES5 output?

**Comment by piotrwitek** [comment]
@omidkrad yes `"target": "es5"`, then I'm running `regenerator -r app.ts-es5.js > app.es5.js`, I needed this mainly because of IE10+ support and it's working fine, just need to include ES6 shim for Promises

**Comment by vigie** [comment]
@piotrwitek I'm missing something - how do you get tsc to compile your code to es5 in the first place if you have async/await syntax in there?

**Comment by piotrwitek** [comment]
@vigie When using compiler you got async compilation error `(Async functions are only available when targeting ECMAScript 6 and higher.),` but transpiled code using generator functions is still emitted just fine, in Atom I'm using target es6 for eslint and typescript compile but with noEmit flag so normally I don't see any of these errors during development, then down my workflow I'm using SystemJS/JSPM to build bundle with target es5, on bundle step I don't have to care anymore about compiler errors

**Comment by vigie** [comment]
@piotrwitek A very interesting technique, thanks for taking the time to explain it. I'm currently evaluating introducing Babel into the chain, but its great to know you have had success with this lighter weight option.

**Comment by aleksey-bykov** [comment]
it's late may and still no promises

**Comment by jonaskello** [comment]
Big +1 on this feature. With this feature, my team's development flow could be 100% integrated inside VS2015. Currently, because we need generator functions, we need to introduce double compiling with babel and loose a lot of VS2015 integration (compile on save, error reporting, and debugging). Really looking forward to using this feature once it is implemented.

**Comment by AppShipIt** [comment]
I'm currently using `async` in ES5 using the compiler service as well. However arrow syntax transpiles missing a closure described in https://github.com/Microsoft/TypeScript/issues/7706#issuecomment-227908547. Would it be worth opening a separate issue as it looks like low hanging fruit?

**Comment by MattiasBuelens** [comment]
Is this still on track for TypeScript 2.0? The implementing PR #9175 for this depends on the new emitter #5595, but the milestone of the latter has been updated to 2.1.

I'd appreciate it if one of the core contributors could update the [roadmap](https://github.com/Microsoft/TypeScript/wiki/Roadmap) and the issues linked in the roadmap with the correct milestones. It'd help get a better idea what to expect in 2.0, because right now there's a lot of conflicting information.

**Comment by WanderWang** [comment]
2.0 or 2.1 ?

**Comment by vbcodec** [comment]
Do not release TS 2.0 without async / await for ES5/3, the main feature. Otherwise it will be ridiculous.

**Comment by glen-84** [comment]
@vbcodec I want it just as much as you do, but there is lots of new stuff in 2.0, so it wouldn't be "ridiculous" – maybe just disappointing.

**Comment by AdamWillden** [comment]
It'd just be nice to have some indication from 'those in the know' what is happening with regards to this feature. With respect to varying levels of `disappointing`, personally losing this feature would ruin my ideals of what 2.0 is. 

For the mean time curiosity and not knowing is killing me more than the prospect of not having this feature :smiling_imp:

**Comment by niieani** [comment]
I would be very sad if this doesn't make it into 2.0, but it wouldn't be the end of the world -- I can just keep the Babel step a few more months, e.g. until 2.1 arrives with it. I'll just be glad to have 2.0 finally shipped, cause it brings so much other goodness for which there are no any workarounds that our community ([Aurelia Framework](http://aurelia.io)) is waiting for. 
A 2nd step Babel transpilation can still be used to get _async/await_ with a `es5` target.

**Comment by vbcodec** [comment]
@glen-84 
The fact is that implementing async/await is alerady ridiculous. It is few years when they 'plan to implement it in next version'. Maybe they do not want to implement it at all, and just playing bad game with us ? Only IE 11- need this feature, because Chrome and FF have support to ES6, and cost of implementing is very high. Why to make serious investment for leagcy standard on legace browser on legacy Windows (8.1-) ? Just use Babel of lambdas if you want support legacy tech. This may be line of thinking by TS team.

The 'new stuff' is null-check, but this is much less usefull than anybody wish. Preventing null exceptions lead to complex declarations on code and data structures. JS/TS code is very dynamic, most libraries aren't designed with non-null philosophy, and will cayse leaks and serious resistance while bending to non-null behaviour. That's why most devs will stay with 'more risky but much simpler code'.

**Comment by niieani** [comment]
@vbcodec there is much more 'new stuff' coming with 2.0 than just the _null-check_. For me, type declarations on `this` of any function is much more important, just as granular `lib` selection or the new typing resolution strategies are.

In any case, I would suggest not to flame this topic with irrelevant comments further, let's all patiently wait for an official response.

**Comment by mhegazy** [comment]
As noted in [today's release announcement](https://blogs.msdn.microsoft.com/typescript/2016/07/11/announcing-typescript-2-0-beta/), we are happy with the implementation we have today for async-await (see https://github.com/Microsoft/TypeScript/pull/9175). The PR is in final code review iterations; and we hope it will be merged in master within few weeks. once that is done, it should be available for the general public in the nightly releases (`typescript@next`).

**Comment by AdamWillden** [comment]
We've had our official response - 2.1 it is.

**Comment by jamespacileo** [comment]
@mhegazy any news? is the merge imminent? Really looking forward to this feature release! :+1: :smile:

**Comment by HoosierMike** [comment]
Yes, we desperately need async/await for ES5 ...

**Comment by jamespacileo** [comment]
@HoosierMike desperate might be too much, but it does indeed feel like it! It overjoys me already to use async/await in ES6 node, can't wait to use it for ES5 frontend apps.

**Comment by rybolt** [comment]
I see it's on the roadmap for version 2.1 , but how to know if it is in nightly build yet ?  I'm assuming that is when we can easily grab at the bits.  Thanks

**Comment by devzsolt** [comment]
Any ETA perhaps?

**Comment by glen-84** [comment]
@devzsolt https://github.com/Microsoft/TypeScript/pull/9175#issuecomment-244181037

**Comment by devzsolt** [comment]
great news, thanks @glen-84!

**Comment by rbuckton** [comment]
Down-level generator support for async functions (async/await) is now in master and should be available via `typescript@next` tomorrow.

**Comment by jonaskello** [comment]
@rbuckton  Will this support also cover compiling ES6 generator functions (function*/yield) to ES5? I'm guessing it does but from the description it seems to be only about async/await.

**Comment by MattiasBuelens** [comment]
@jonaskello I saw that there was some work on partially down-leveling generator functions (not just async/await) in #10106, but that was apparently closed because the restrictions on what is and is not supported weren't clear (see #10307).

I think the idea is to implement just enough to transpile async/await, but not yet transpile *all* generator functions. That way, the TypeScript team can ship this feature sooner because it only needs partial support for generators. I believe full generator support is still somewhere on the roadmap, but simply for a later version. (Don't quote me on that though, I'm not part of the TypeScript team. I just follow their progress in some interesting PRs. :stuck_out_tongue: )

**Comment by jonaskello** [comment]
If that is true then what has been merged into master does not solve this issue (which is titled ES6 Generator support). I had really hoped for generator function support. Maybe I am in the minority here but I have no use for async/await. I do all my side-effects in redux-saga which only uses generator functions. One could argue that support for features that are part of ES6 (generator functions) should be have priority over features that will (maybe) be out in a later ES version (async/await). But still nice to see that progress is made in areas related to generator functions!

**Comment by Kovensky** [comment]
Nit: async/await won't "maybe" be out, they are stage-4, which means they
_are_ standard. The annual spec publishing is a formality.

**Comment by jonaskello** [comment]
@Kovensky  Sorry about that, I don't follow async/await that closely. Last I checked it was stage-3 but things move quickly :-). Updated my comment accordingly.

**Comment by kenhahn85** [comment]
I'm still getting `error TS1311: Async functions are only available when targeting ECMAScript 2015 or higher.` when compiling async-await functions with today's `typescript@next`. Is there a compiler flag that needs to be set?

**Comment by hermitdemschoenenleben** [comment]
Thank you for your work! However, with today's version I still get the error:

TypeScript error: app/app.ts(32,3): Error TS1311: Async functions are only available when targeting ECMAScript 6 and higher.


> require('typescript').version
'2.1.0-dev.20160909'

**Comment by jotaoncode** [comment]
It would be great to use redux-saga with typescript and es6 generators.

http://yelouafi.github.io/redux-saga/

**Comment by drywolf** [comment]
@jotaoncode You could already do that if you want use webpack+babel+typescript for your build process, like so:

_**webpack.config.js**_

    module: {
        loaders: [
            {
                test: /\.tsx?$/,
                loader: 'babel-loader!awesome-typescript-loader',
                // ...
            },
        ]
    },

**_.babelrc_**

    {
        "presets": ["es2015", "stage-0"],
        "plugins": ["transform-regenerator"]
    }

This way you can use ES2015 Generators & Async/Await syntax in combination with redux-saga, and down-level compile the code to ES5 (which would then also run in IE11)

**Comment by jonaskello** [comment]
@jotaoncode Seems like it is not going to happen, see #3975.

**Comment by jotaoncode** [comment]
@drywolf I saw your repo related to typescript webpack generator sample, also seems like they already fixed the problem related to awesome-typescript-loader, so probably the temporary fix you did is no longer needed.
Many thanks for the help :)

**Comment by kdalgaard** [comment]
I've been waiting for support for down-level generators in version 1.5, then 1.6, then 2.0, then 2.1.... Now I'm waiting again to hear from the team when they expect it to come. So dissatisfying...

My use case is `redux-saga` for side-effects and long-lived transactions.

**Comment by jocull** [comment]
I believe this is on the roadmap, is already merged, and is slated for November 2016 under 2.1. You can probably use these things today by using `typescript@next`?
https://github.com/Microsoft/TypeScript/wiki/Roadmap
https://github.com/Microsoft/TypeScript/pull/9175

My apologies if I'm missing the point here entirely. It's my understanding that async functions and generators are the same under the hood, but it may only be a subset of generator features that are implemented.

**Comment by yortus** [comment]
@jocull downlevel generator support has been removed from the roadmap altogether and looks unlikely to be implemented at all - see https://github.com/Microsoft/TypeScript/issues/3975#issuecomment-250859415.

Given that, I'm not sure why this issue is still open or what it is still tracking.

**Comment by aharpervc** [comment]
@kdalgaard 
> I've been waiting for support for down-level generators in version 1.5, then 1.6, then 2.0, then 2.1.... Now I'm waiting again to hear from the team when they expect it to come. So dissatisfying...

Same. Been waiting for this for a long time. And now it's silently not happening? Lame.

**Comment by jonaskello** [comment]
I guess for the most part you can develop using generator functions with ES6 output and use Chrome which has native support for generator functions. We have an interesting corner case though. We have a legacy Silverlight application that we are converting to react with redux-saga. So we need to run both SIlverlight and redux-saga on the same page. Silverlight does not run in Chrome, Edge so we cannot use them. Internet Explorer runs Silverlight but does not support generator functions, and since TS does not downcompile to ES5 we cannot use IE. Fortunately for us Firefox supports both SIlverlight and generator functions, however it does not support debugging from Visual Studio.

**Comment by glen-84** [comment]
@jonaskello Would [regenerator](https://github.com/facebook/regenerator) work? Or does TS do something to the generators when emitting ES5?

**Comment by piotrwitek** [comment]
Hi @glen-84, Regenerator would work, you can read explanation and check working solution in my starter-kit project here: https://github.com/piotrwitek/react-redux-typescript-starter-kit#asyncawaitgenerators-transformation-when-targeting-es3es5-without-babel

**Comment by jonaskello** [comment]
@glen-84 Yupp it would work but then you need to include something like gulp or webpack or babel or something else that multiplies the complexity of the setup. It would be nice to just code Typescript in Visual Studio with full ES2015 support out of the box. Just code some TypeScript, set a breakpoint, hit F5 and off you go. Especially since we have a team that has been doing that for C#/Silverlight. Granted in the end you need bundling for production but it would be nice if not every developer has to learn about these tools.

**Comment by glen-84** [comment]
I wonder if TypeScript could just use regenerator internally? I guess they might want to avoid dependencies though.

**Comment by RyanCavanaugh** [comment]
@rbuckton can you provide a summary of where we are with this?

**Comment by kdalgaard** [comment]
@piotrwitek I have read that paragraph of your starter readme over and over, and I still can't see how you would perform the first step, which is compiling TypeScript down to ES5 with generators intact. I believe it will just throw an error when seeing this syntax?

In your starter repo, the compile target for TypeScript is ES6, which won't help at all – Babel will still be needed to transpile it the rest of the way down.

If there exists some way to make TypeScript ignore generators syntax when compiling to ES5, I haven't found it yet 👎

**Comment by piotrwitek** [comment]
@kdalgaard thanks for pointing that out it means I'll have to add a new paragraph clarifying that which need you to know how plugin-typescript works.

To put it simply there is a seperate TS compiler config for production in jspm.config.js which target ES5. TS will normally emit valid ES5 code with generators passed through thats where regenerator kicks in to finish the job. Now you have valid ES5 with generator functions transpiled.

For development I use only ES6 because you will use evergreen browser anyway for dev so they have native generator support, no need to downlevel that during development and complicate things.
Then you will use build script to create ES5 code bundle using ES5 config explained above.

**Comment by kdalgaard** [comment]
@piotrwitek I wish it were possible, what you describe, but on my machine, trying to compile a TS script with generators in it to ES5 results in errors in the console, saying that this syntax is not supported for ES5 targets.

**Comment by MaximBalaganskiy** [comment]
You can transpile ts to es6 while editing and configure jspm for babel es6
-> es5 transpilation

**Comment by kdalgaard** [comment]
@MaximBalaganskiy The point in question is how to avoid Babel.

**Comment by MaximBalaganskiy** [comment]
Oh, I see. Then Piotr's approach should work. You configure tsconfig for
es6 and jspm for es5. Thus no errors in a dev console and right code in
browser.

**Comment by kdalgaard** [comment]
@MaximBalaganskiy I completely fail to see where in that description ES6 is compiled to ES5 without the use of Babel.

**Comment by jonaskello** [comment]
@kdalgaard I think the confusion here might be that when you use `jspm` then you do not use `tsc` at all. Instead something called `plugin-typescript` is used and it handles the compiling by calling into typescript API. So I guess `plugin-typescript` just ignores generator functions and outputs them. Then regenerator can run and fix them. If you look as [here](https://github.com/piotrwitek/react-redux-typescript-starter-kit/blob/dc86c62eafae432bccc243300146b82d936b262e/jspm.config.js#L131) in @piotrwitek's starter you will see that it actually compiles to es5, but not with `tsc`, instead `plugin-typescript` is used as you can see [here](https://github.com/piotrwitek/react-redux-typescript-starter-kit/blob/dc86c62eafae432bccc243300146b82d936b262e/jspm.config.js#L128).

**Comment by kdalgaard** [comment]
@jonaskello Thanks for that.

I wonder how `plugin-typescript` is able to use the TypeScript compiler in a way that keeps generators intact. At any rate, it seems like a pretty flaky strategy from `plugin-typescript` to just output ES6 syntax even if the target is set to ES5...

**Comment by piotrwitek** [comment]
@kdalgaard @jonaskello it is working because TypeScript compiler can emit files even when there are errors, so it will emit code compiled to es5 without transpiling generators, but the rest of transformation is good

**Comment by kdalgaard** [comment]
@piotrwitek I understand as much. The million dollar question is: How?

I'm looking for a more TypeScript generic method than the somewhat questionable programmatic use in `plugin-typescript` for JSPM, where errors are apparently ignored.

**Comment by jonaskello** [comment]
@kdalgaard I have not tried it but I think the `tsc` option is --noEmitOnError = false, which is the default. I think even if `tsc` will yell at you and exit with error it will still emit what it can. I guess `plugin-typescript` does the same but without yelling at you or exiting with error.

**Comment by kdalgaard** [comment]
@jonaskello Can't seem to get it to work. Maybe something with my Webpack setup or trouble with TSLint...

How do you all become aware of actual script errors, if everything is output anyway?

**Comment by jonaskello** [comment]
@kdalgaard I can describe my workflow if it helps. In my previous project I used webpack. Everything was compiled, type-checked and bundled in-memory using webpacks dev-server, and then it hit the browser. So when it hit the browser I was sure it compiled and was type-checked. In the beginning this was really fast. Now that project has grown, initial webpack build takes 36263ms and incremental takes 16981ms. So for my current project I decided to try jspm instead. Not sure if it is any better, but at least it is different :-). So I found @piotrwitek's jspm starter and did not really understand how that worked because it is so different from the webpack flow. Then I finally understood. You have to realise that code is *not* compiled, *not* type-checked and *not* bundled before hitting the browser. It is simply just transpiled, bascially it just strips the type information without caring about it. This makes it possible to have JS hitting the browser fast which makes it possible to make changes in TS and iterate really fast. But I had the same question as you, how do I know if there is a compile error? Well, you have to have a second place where you do that. For example you can use atom with the plugin that incrementially compiles and type-checks. Or you can simply use any editor and just run tsc --watch in the termial, which is what I do. So what I finally realised was that compiling/type-checking could be a separate concern to getting the JS into the browser. Granted this is very different from how webpack works and you have unlearn a bit. And of course this is not the only workflow for TS in jspm, it is just the one I have currently adopted.

**Comment by kdalgaard** [comment]
@jonaskello Thanks for the explanation. I would expect Webpack to continue being fast due to hot module replacement, but I don't know...

**Comment by sheerun** [comment]
@DanielRosenwasser @jonaskello As demonstrated by [fast-async](https://www.npmjs.com/package/fast-async) you don't need generators to support async/await, **promises are enough**. Please don't despair and implement async/await for es5 in typescript.. The the only thing that is blocking me to switch.

**Comment by jonaskello** [comment]
@sheerun This issue it not about async/await. It is about generator functions. AFAIK async/await -> ES5 is already implemented in ts@next.

**Comment by kdalgaard** [comment]
@sheerun What @jonaskello said.

**Comment by jonaskello** [comment]
Now that #5595 has landed, is there any chance this issue can make it back onto the roadmap?

**Comment by wallverb** [comment]
Looks like there is a chance: https://github.com/Microsoft/TypeScript/pull/12346
